rama-tls-rustls/src/client/connector_data.rs

```diff
@@ -0,0 +1,256 @@
+use crate::dep::pki_types::{CertificateDer, PrivateKeyDer, PrivatePkcs8KeyDer};
+use crate::dep::rcgen::{self, KeyPair};
+use crate::dep::rustls::RootCertStore;
+use crate::dep::rustls::{ALL_VERSIONS, ClientConfig};
+use crate::key_log::KeyLogFile;
+use crate::verify::NoServerCertVerifier;
+use rama_core::error::{BoxError, ErrorContext, OpaqueError};
+use rama_net::address::Host;
+use rama_net::tls::{ApplicationProtocol, KeyLogIntent};
+use rustls::client::danger::ServerCertVerifier;
+use std::sync::{Arc, OnceLock};
+
+#[derive(Debug, Clone)]
+/// Internal data used as configuration/input for the [`super::HttpsConnector`].
+///
+/// Created by converting a [`rustls::ClientConfig`] into it directly,
+/// or by using [`TlsConnectorDataBuilder`] to build this in a more ergonomic way.
+pub struct TlsConnectorData {
+    pub client_config: Arc<ClientConfig>,
+    pub server_name: Option<Host>,
+    pub store_server_certificate_chain: bool,
+}
+
+impl From<ClientConfig> for TlsConnectorData {
+    #[inline]
+    fn from(value: ClientConfig) -> Self {
+        Arc::new(value).into()
+    }
+}
+
+impl From<Arc<ClientConfig>> for TlsConnectorData {
+    fn from(value: Arc<ClientConfig>) -> Self {
+        Self {
+            client_config: value,
+            server_name: None,
+            store_server_certificate_chain: false,
+        }
+    }
+}
+
+impl TlsConnectorData {
+    /// Create a default [`TlsConnectorData`] that is focussed
+    /// on providing auto http connections, meaning supporting
+    /// the http connections which `rama` supports out of the box.
+    pub fn new_http_auto() -> Result<TlsConnectorData, OpaqueError> {
+        Ok(TlsConnectorDataBuilder::new()
+            .with_env_key_logger()?
+            .with_http_versions(&[ApplicationProtocol::HTTP_11, ApplicationProtocol::HTTP_2])
+            .build())
+    }
+
+    /// Create a default [`TlsConnectorData`] that is focussed
+    /// on providing http/1.1 connections.
+    pub fn new_http_1() -> Result<TlsConnectorData, OpaqueError> {
+        Ok(TlsConnectorDataBuilder::new()
+            .with_env_key_logger()?
+            .with_http_versions(&[ApplicationProtocol::HTTP_11])
+            .build())
+    }
+
+    /// Create a default [`TlsConnectorData`] that is focussed
+    /// on providing h2 connections.
+    pub fn new_http_2() -> Result<TlsConnectorData, OpaqueError> {
+        Ok(TlsConnectorDataBuilder::new()
+            .with_env_key_logger()?
+            .with_http_versions(&[ApplicationProtocol::HTTP_2])
+            .build())
+    }
+}
+
+/// [`ClientConfigBuilder`] can be used to construct [`rustls::ClientConfig`] for most common use cases in Rama.
```

Comment:
Try to keep the first docstring line short as markdown being markdown a new line is only an actual newline if it is separated by an empty newline. So the problem with things like this is that your summary (first line) of `TlsConnectorDataBuilder` is going to be very long (in stuff like docs.rs)